### PR TITLE
feat(css): implement @layer and @import rules

### PR DESCRIPTION
**Pull request by noreiller**
The changes add the support of `@layer` css at-rule and implements the rules when using `@import` CSS at-rule.

It is part of CSS experiment.

**`@layer`**
The following code would be handled by webpack, especially in CSS modules.
```css
@layer base, special;

body {
  background: green;
  font-family: "Open Sans";
}

@layer special {
  .item {
    color: rebeccapurple;
  }
}

@layer base {
  .item {
    color: black;
    border: 5px solid black;
    font-size: 1.3em;
    padding: .5em;
  }
}
```

**`@import` rules**
This code:
```css
/* style.css */
@import url( "style2.css" ) layer( base ) supports( font-weight: bold ) screen and (min-width: 1024px);

/* style2.css */
body {
  font-weight: bold;
  text-decoration: underline;
}
```
would be transformed to:
```css
@layer base {
  @supports(font-weight: bold) {
    @media screen and (min-width: 1024px) {
      body {
        font-weight: bold;
        text-decoration: underline;
      }
    }
  }
}
```

**Tests**
Some tests have been added for CSS modules.
The ones dedicated to `@import` rules have also been added but, unfortunately, JSDOM (which relies on CSSOM) does not handle well at-rules so they are disabled until the support is updated.

### CODE REVIEW

File: lib/dependencies/CssImportDependency.js [hunk 1]
```diff
@@ -64,6 +68,28 @@ CssImportDependency.Template = class CssImportDependencyTemplate extends (
 		const dep = /** @type {CssImportDependency} */ (dependency);
 
 		source.replace(dep.range[0], dep.range[1] - 1, "");
+
+		if (
+			// Do nothing more for external modules
+			/^(\/\/|https?:\/\/|std:)/.test(dep.request) ||
```
Comment (alexander-akait) [inline lib/dependencies/CssImportDependency.js:74]: We can handle externals styles

Comment (noreiller) [reply lib/dependencies/CssImportDependency.js:74]: External styles are already considered as externals assets and so, no need to transform them. It's done there: https://github.com/webpack/webpack/blob/main/lib/WebpackOptionsApply.js#L119-L163.

Comment (alexander-akait) [reply lib/dependencies/CssImportDependency.js:74]: I mean https://webpack.js.org/configuration/experiments/#experimentsbuildhttp, we can bundle `http`/`https` CSS

Comment (noreiller) [reply lib/dependencies/CssImportDependency.js:74]: Oh yes, I forgot that. Thanks, I will check and update the PR.

Comment (noreiller) [reply lib/dependencies/CssImportDependency.js]: @alexander-akait The code is updated with `buildHttp` experiment and also with assets replacements.

**Review by alexander-akait — APPROVED** [review]
Looks good to me

**Review by sokra — CHANGES_REQUESTED** [review]

File: lib/dependencies/CssImportDecoratorDependency.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	.reduce(tpl => Template.indent(tpl), replacedSource)}
+${decorations
+	.map(([before, after], idx) => {
+		return Array(decorations.length - 1 - idx)
+			.fill()
+			.reduce(tpl => Template.indent(tpl), after);
+	})
+	.join("\n")}`
+		);
+	}
```
[the remark this answers is not in the document]
Comment (noreiller) [reply lib/dependencies/CssImportDecoratorDependency.js:103]: I did it aft first try but I add issues with source mapping. I think it's related to the issue you mentioned in another comment.

File: lib/css/walkCssTokens.js [hunk 3]
```diff
@@ -357,6 +361,119 @@ const consumePotentialUrl = (input, pos, callbacks) => {
...
+				}
+			} else if (cc === CC_LEFT_PARENTHESIS) {
+				level += 1;
+				pos++;
+			} else {
+				pos++;
+			}
+			if (pos === input.length) return pos;
+			cc = input.charCodeAt(pos);
+		}
```
[the remark this answers is not in the document]
Comment (noreiller) [reply lib/css/walkCssTokens.js:468]: Yes, I will refactor that.

**Comment by noreiller** [comment]
@sokra Thanks for the review, I will change the proposal according to what you asked but it won't be until a few weeks since I have some holidays.
In the mean time, if you can check other PRs related to Webpack CSS (#15812 #15818), I would be really thankful.

File: lib/css/walkCssTokens.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (noreiller) [reply lib/css/walkCssTokens.js]: Done.

**Comment by noreiller** [comment]
Hello @sokra @alexander-akait, will you continue to implement the webpack experimental CSS feature or accept new PRs?

**Review by alexander-akait — APPROVED** [review]

**Comment by alexander-akait** [comment]
@noreiller Let's do rebase, thank you

**Comment by alexander-akait** [comment]
Thank you I will review soon

**Comment by noreiller** [comment]
> After merge - #16978 (because it fixed some bugs in tokenizer), I will focus on it, I have idea how we can simplify it

It's gonna be another complicated rebase... :open_mouth: 
I'll wait for your merge and then take care of it.

**Comment by alexander-akait** [comment]
@noreiller don't worry, you have

>  Maintainers are allowed to edit this pull request.

So I will do rebase

**Comment by noreiller** [comment]
> @noreiller don't worry, you have
> 
> > Maintainers are allowed to edit this pull request.
> 
> So I will do rebase

Ok, thanks :pray:

**Pull request closed without merging**